### PR TITLE
`fn sgr_{5x5,3x5,mix}`: Deduplicate w/ generics

### PR DESCRIPTION
**Pull request by randomPoison**

### CODE REVIEW

File: src/looprestoration.rs [hunk 1]
```diff
@@ -726,3 +726,223 @@ pub(crate) unsafe extern "C" fn selfguided_filter<BD: BitDepth>(
         }
     };
 }
+
+// TODO(randomPoison): Temporarily pub until init logic is deduplicated.
+pub(crate) unsafe extern "C" fn sgr_5x5_c<BD: BitDepth>(
```
Comment (kkysen) [inline src/looprestoration.rs:731]:
```suggestion
pub(crate) unsafe extern "C" fn sgr_5x5_c_erased<BD: BitDepth>(
```

Did you want to mark them erased, too?

File: src/looprestoration.rs [hunk 2]
```diff
@@ -726,3 +726,223 @@ pub(crate) unsafe extern "C" fn selfguided_filter<BD: BitDepth>(
...
+    mut p: *mut libc::c_void,
+    stride: ptrdiff_t,
+    left: *const libc::c_void,
+    mut lpf: *const libc::c_void,
```
Comment (kkysen) [inline src/looprestoration.rs:735]:
```suggestion
    p: *mut libc::c_void,
    stride: ptrdiff_t,
    left: *const libc::c_void,
    lpf: *const libc::c_void,
```

Same unused `mut`s on args in the other `fn`s, too.

File: src/looprestoration.rs [hunk 3]
```diff
@@ -726,3 +726,223 @@ pub(crate) unsafe extern "C" fn selfguided_filter<BD: BitDepth>(
...
+        edges,
+        BD::from_c(bitdepth_max),
+    )
+}
+
+unsafe fn sgr_5x5_rust<BD: BitDepth>(
+    mut p: *mut BD::Pixel,
+    stride: ptrdiff_t,
+    left: *const [BD::Pixel; 4],
+    mut lpf: *const BD::Pixel,
```
Comment (kkysen) [inline src/looprestoration.rs:759]:
```suggestion
    lpf: *const BD::Pixel,
```

Same for the other `*_rust` `fn`s.

File: src/looprestoration.rs [hunk 4]
```diff
@@ -726,3 +726,223 @@ pub(crate) unsafe extern "C" fn selfguided_filter<BD: BitDepth>(
...
+        while i < w {
+            let v = w0 * dst0[(j * 384 + i) as usize].as_::<libc::c_int>()
+                + w1 * dst1[(j * 384 + i) as usize].as_::<libc::c_int>();
+            *p.offset(i as isize) = iclip(
+                (*p.offset(i as isize)).as_::<libc::c_int>()
+                    + (v + ((1 as libc::c_int) << 10) >> 11),
+                0 as libc::c_int,
+                bd.bitdepth_max().as_(),
+            )
+            .as_();
```
Comment (kkysen) [inline src/looprestoration.rs:942]:
```suggestion
            *p.offset(i as isize) = bd.iclip_pixel(
                (*p.offset(i as isize)).as_::<libc::c_int>()
                    + (v + ((1 as libc::c_int) << 10) >> 11),
            );
```

I think it should be this?  Double-check, but it needs to be `bd.iclip_pixel(...)`.

File: src/looprestoration_tmpl_16.rs [hunk 5]
```diff
@@ -130,15 +129,19 @@ extern "C" {
...
+cfg_if! {
+    if #[cfg(any(target_arch = "arm", target_arch = "aarch64"))] {
+        use crate::src::looprestoration::LrEdgeFlags;
+        use crate::src::looprestoration::LooprestorationParams;
+    }
+}
+
 use crate::src::looprestoration::wiener_erased;
 use crate::src::looprestoration::Dav1dLoopRestorationDSPContext;
+use crate::src::looprestoration::{sgr_3x3_c, sgr_5x5_c, sgr_mix_c};
```
Comment (kkysen) [inline src/looprestoration_tmpl_16.rs:144]: Can you split these up?

File: src/looprestoration_tmpl_16.rs [hunk 6]
```diff
@@ -130,15 +129,19 @@ extern "C" {
...
-use crate::include::common::intops::iclip;
-use crate::src::looprestoration::padding;
-use crate::src::looprestoration::selfguided_filter;
+#[cfg(feature = "asm")]
+cfg_if! {
+    if #[cfg(any(target_arch = "arm", target_arch = "aarch64"))] {
+        use crate::src::looprestoration::LrEdgeFlags;
+        use crate::src::looprestoration::LooprestorationParams;
+    }
+}
```
Comment (kkysen) [inline src/looprestoration_tmpl_16.rs:140]: Why's the `#[cfg]` logic done twice here and with `cfg_if!`?

Can it be this?

```suggestion
#[cfg(all(feature = "asm", any(target_arch = "arm", target_arch = "aarch64")))]
use crate::{
    src::looprestoration::LrEdgeFlags,
    src::looprestoration::LooprestorationParams,
};
```

File: src/looprestoration_tmpl_8.rs [hunk 7]
```diff
@@ -127,212 +125,17 @@ extern "C" {
...
-                *p.offset(i as isize) as libc::c_int + (v + ((1 as libc::c_int) << 10) >> 11),
-            ) as pixel;
-            i += 1;
-        }
-        p = p.offset(stride as isize);
-        j += 1;
+#[cfg(feature = "asm")]
+cfg_if! {
+    if #[cfg(any(target_arch = "arm", target_arch = "aarch64"))] {
+        use crate::include::stddef::*;
```
Comment (kkysen) [inline src/looprestoration_tmpl_8.rs:136]:
```suggestion
        use crate::include::stddef::ptrdiff_t;
```

File: src/looprestoration_tmpl_16.rs (hunk 5 above)
Comment (kkysen) [reply src/looprestoration_tmpl_16.rs:144]: Same for 8bpc.

File: src/looprestoration_tmpl_16.rs (hunk 6 above)
Comment (kkysen) [reply src/looprestoration_tmpl_16.rs:140]: Same for 8bpc.

File: src/looprestoration.rs [hunk 8]
```diff
@@ -726,3 +726,223 @@ pub(crate) unsafe extern "C" fn selfguided_filter<BD: BitDepth>(
...
+        let mut i = 0;
+        while i < w {
+            let v = w0 * dst[(j * 384 + i) as usize].as_::<libc::c_int>();
+            *p.offset(i as isize) = iclip(
+                (*p.offset(i as isize)).as_::<libc::c_int>()
+                    + (v + ((1 as libc::c_int) << 10) >> 11),
+                0 as libc::c_int,
+                bd.bitdepth_max().as_(),
+            )
+            .as_();
```
Comment (kkysen) [inline src/looprestoration.rs:791]: `bd.iclip_pixel(...)`.

**Review by kkysen — CHANGES_REQUESTED** [review]
The `bd.iclip_pixel(...)`s are missing/bitdepth-specific, and there are some other smaller comments as well.

File: src/looprestoration.rs (hunk 1 above)
Comment (randomPoison) [reply src/looprestoration.rs:728]: Ah yeah, let's keep the naming convention consistent.

File: src/looprestoration_tmpl_16.rs (hunk 6 above)
Comment (randomPoison) [reply src/looprestoration_tmpl_16.rs:140]: Ah yeah, I like your suggested approach better. The only thing is that rustfmt tries to put all the imports on the same line, so I also threw a `#[rustfmt::skip]` on there. I don't love having to do that, but these conditional imports are just to suppress unused warnings, and will go away once deduplication for `looprestoration.rs` is done.

File: src/looprestoration.rs (hunk 2 above)
Comment (randomPoison) [reply src/looprestoration.rs]: Removed.

File: src/looprestoration.rs (hunk 4 above)
Comment (randomPoison) [reply src/looprestoration.rs]: Done.

File: src/looprestoration_tmpl_16.rs (hunk 5 above)
Comment (randomPoison) [reply src/looprestoration_tmpl_16.rs]: Done.

File: src/looprestoration.rs [hunk 9]
```diff
@@ -726,3 +726,223 @@ pub(crate) unsafe extern "C" fn selfguided_filter<BD: BitDepth>(
...
+        let mut i = 0;
+        while i < w {
+            let v = w1 * dst[(j * 384 + i) as usize].as_::<libc::c_int>();
+            *p.offset(i as isize) = iclip(
+                (*p.offset(i as isize)).as_::<libc::c_int>()
+                    + (v + ((1 as libc::c_int) << 10) >> 11),
+                0 as libc::c_int,
+                bd.bitdepth_max().as_(),
+            )
+            .as_();
```
[the remark this answers is not in the document]
Comment (randomPoison) [reply src/looprestoration.rs]: Done.

File: src/looprestoration.rs (hunk 8 above)
Comment (randomPoison) [reply src/looprestoration.rs]: Done.

File: src/looprestoration_tmpl_16.rs (hunk 6 above)
Comment (kkysen) [reply src/looprestoration_tmpl_16.rs]: That sounds good.  `#[rustfmt::skip]` should be fine for temporary things.

**Pull request merged**